scadnano/scadnano.py

```diff
@@ -3895,51 +3893,79 @@ def from_scadnano_json_map(
             helices.append(helix)
             idx_default += 1
+        return (helices, group_to_pitch_yaw, pitch_yaw_to_helices)
 
-        # helix groups
+    @staticmethod
+    def _groups_and_grid_from_json(json_map: dict, helices: List[Helix], group_to_pitch_yaw:  Dict[str, Tuple[float, float, int]], pitch_yaw_to_helices:  Dict[Tuple[float, float], List[Helix]]) -> Tuple[Dict[str, HelixGroup], Grid]:
+        """Returns map of helix group names to group as well as the grid
+
+        If multiple helix groups are used, then groups pitch and yaw will be the
+        pitch and yaw given in the json map plus the pitch and yaw values
+        provided in group_to_pitch_yaw
+
+        If a single helix group is used, then new helix groups will be created using
+        the provided group_to_pitch_yaw map. Because new helix groups are created,
+        helices will be modfied so that each helix's group field is set to the new
+        group it has been assigned to.
+
+        In most cases, grid will simply be what is given in the json_map.
+        There is a special case where if the default helix group is used,
+        then the grid may be initially set to some value. If individual helix
+        pitch and rolls were set, then new helix groups are created, meaning
+        the grid field is no longer valid.
+        """
+        # Return values
         groups = None
+        grid = optional_field(None, json_map, grid_key)
+
+        # Get helix groups provided from json_map
+        multiple_groups_used = Design._num_helix_groups(json_map) > 1
+        single_group_used = not multiple_groups_used
         if groups_key in json_map:
             groups_json = json_map[groups_key]
             groups = {}
             for name, group_json in groups_json.items():
                 num_helices_in_group = sum(1 for helix in helices if helix.group == name)
                 groups[name] = HelixGroup.from_json(group_json, num_helices=num_helices_in_group)
 
+        # Add individual helix pitch and yaw
         if multiple_groups_used:
             for (group_name, (pitch, yaw, _)) in group_to_pitch_yaw.items():
                 groups[group_name].pitch += pitch
-                groups[group_name].yaw += pitch
+                groups[group_name].yaw += yaw
 
+        # Add new helix groups if individual helix pitch and yaw set
         if single_group_used:
             # New groups to add
             new_groups = {}
             # The only group (take into account case when default helix group is used)
             group = list(groups.values())[0] if groups is not None else HelixGroup()
 
-            for ((integerized_pitch, integerized_yaw), helix_list) in pitch_yaw_to_helices.items():
-                new_pitch = group.pitch + integerized_pitch * floating_point_tolerance
-                new_yaw = group.yaw + integerized_yaw * floating_point_tolerance
+            for ((helix_pitch, helix_yaw), helix_list) in pitch_yaw_to_helices.items():

```

Comment:
Put in some comments describing which portions of this code we expect to be only needed for backwards compatibility. It's a mix between code that will always be needed for import in v0.16.0 JSON, and code that will only be executed when encountering previous versions (e.g., v0.15.0) that stored pitch and yaw in Helix. Make sure it's easy to tell the difference between those two kinds of code by adding some comments saying when something is there for backward compatibility.